File: src/Algorithms/bfs.js
export function bfs(grid, startNode, endNode) {
    const visitedNodesInOrder = [];
    const queue = [];
    startNode.distance = 0;
    queue.push(startNode);
    while (queue.length !== 0) {
        const currentNode = queue.shift();
        if (currentNode.isWall) continue;
        if (currentNode.distance === Infinity) return visitedNodesInOrder;
        currentNode.isVisited = true;
        visitedNodesInOrder.push(currentNode);
        if (currentNode === endNode) return visitedNodesInOrder;
        const unvisitedNeighbors = getUnvisitedNeighbors(currentNode, grid);
        for (const neighbor of unvisitedNeighbors) {
            neighbor.distance = currentNode.distance + 1;
            neighbor.previousNode = currentNode;
            neighbor.isVisited = true;  // Mark as visited here
            queue.push(neighbor);
        }
    }
}

function getUnvisitedNeighbors(node, grid) {
    const neighbors = [];
    const {col, row} = node;
    if (row > 0) neighbors.push(grid[row - 1][col]);
    if (row < grid.length - 1) neighbors.push(grid[row + 1][col]);
    if (col > 0) neighbors.push(grid[row][col - 1]);
    if (col < grid[0].length - 1) neighbors.push(grid[row][col + 1]);
    return neighbors.filter(neighbor => !neighbor.isVisited);
}

File: golang-src/algorithms/bfs.go
package algorithms

import (
	"math"
	"server/maze"
)

// BFS performs a breadth-first search on the grid
func BFSAlgorithm(grid [][]maze.Node, startNode, endNode *maze.Node) []maze.Node {
	visitedNodesInOrder := []maze.Node{}
	queue := []*maze.Node{}
	startNode.Distance = 0
	queue = append(queue, startNode)

	for len(queue) != 0 {
		currentNode := queue[0]
		queue = queue[1:]

		if currentNode.IsWall {
			continue
		}
		if currentNode.Distance == math.MaxInt32 { // Equivalent to Infinity
			return visitedNodesInOrder
		}
		currentNode.IsVisited = true
		visitedNodesInOrder = append(visitedNodesInOrder, *currentNode)

		if currentNode == endNode {
			return visitedNodesInOrder
		}

		unvisitedNeighbors := getUnvisitedNeighbors(currentNode, grid)
		for _, neighbor := range unvisitedNeighbors {
			neighbor.Distance = currentNode.Distance + 1
			neighbor.PreviousNode = currentNode
			neighbor.PreviousNodeIndex = currentNode.Index
			neighbor.IsVisited = true // Mark as visited here
			queue = append(queue, neighbor)
		}
	}

	return visitedNodesInOrder
}
